interfaces/devdoc/file_util_linux.md

```diff
@@ -0,0 +1,69 @@
+# file_util_linux
+
+## Overview
+
+Linux implementation of the file operation functions.
+
+## Exposed API 
+
+```c
+// Defines removed for space constraints
+
+typedef void *HANDLE;
+typedef void *LPSECURITY_ATTRIBUTES;
+
+MOCKABLE_FUNCTION(, HANDLE, CreateFileA, const char*, lpFileName, unsigned long, dwDesiredAccess, unsigned long, dwShareMode, LPSECURITY_ATTRIBUTES, lpSecurityAttributes, unsigned long, dwCreationDisposition, unsigned long, dwFlagsAndAttributes, HANDLE, hTemplateFile);
+MOCKABLE_FUNCTION(, bool, CloseHandle, HANDLE, handle_input);
+```
+
+### CreateFileA
+
+```c
+MOCKABLE_FUNCTION(, HANDLE, CreateFileA, const char*, lpFileName, unsigned long, dwDesiredAccess, unsigned long, dwShareMode, LPSECURITY_ATTRIBUTES, lpSecurityAttributes, unsigned long, dwCreationDisposition, unsigned long, dwFlagsAndAttributes, HANDLE, hTemplateFile);
+```
+
+`CreateFileA` implements the creating file function in Linux. Uses the `open` function in C for Linux.
+
+**SRS_FILE_UTIL_LINUX_09_001: [** If the `full_file_name` input is either empty or `NULL`, `CreateFileA` shall return an `INVALID_HANDLE_VALUE`. **]**
+
+**SRS_FILE_UTIL_LINUX_09_002: [** `CreateFileA` shall allocate memory for the file handle and on success return a non-`NULL` value.  **]**
+
+**SRS_FILE_UTIL_LINUX_09_003: [** If memory allocation for `result` fails, `CreateFileA` shall return a `NULL` value. **]**
```

Comment:
>`NULL` [](http://example.com/codeflow?start=104&length=6)

Reading these specs, it seems like there are two failure modes: `NULL` and `INVALID_HANDLE_VALUE`. 

The .h file defines `INVALID_HANDLE_VALUE` to be NULL. 

Is that intended? How can a user differentiate between these two error modes? Does the user need to differentiate between them?

Reply:
CreateFileA in windows returns INVALID_HANDE_VALUE for any failures in creating/opening a file handle. So, every time there is an error I will return INVALID_HANDLE_VALUE. I've went back and fixed it so instead of NULL, there is only one failure mode of INVALID_HANDLE_VALUE.